csharp/src/Ice/Reference.cs

```diff
@@ -1524,192 +1318,138 @@ private void ClearConnection(Connection connection)
-                                                                       excludedConnectors,
-                                                                       PreferNonSecure,
-                                                                       cancel).ConfigureAwait(false);
-                                break;
+                                _connection = connection;
                             }
-                            catch (Exception)
-                            {
-                                if (endpoint == lastEndpoint)
-                                {
-                                    throw;
-                                }
-                            }
-                        }
-                    }
-                    Debug.Assert(connection != null);
-
-                    if (RouterInfo != null)
-                    {
-                        await RouterInfo.AddProxyAsync(IObjectPrx.Factory(this)).ConfigureAwait(false);
-
-                        // Set the object adapter for this router (if any) on the new connection, so that callbacks from
-                        // the router can be received over this new connection.
-                        if (RouterInfo.Adapter != null)
-                        {
-                            connection.Adapter = RouterInfo.Adapter;
+                            break;
                         }
-                    }
-                }
-                catch (Exception ex)
-                {
-                    if (LocatorInfo != null && IsIndirect)
-                    {
-                        LocatorInfo.ClearCache(this);
-                    }
-
-                    if (cached)
-                    {
-                        TraceLevels traceLevels = Communicator.TraceLevels;
-                        if (traceLevels.Retry >= 2)
+                        catch
                         {
-                            Communicator.Logger.Trace(TraceLevels.RetryCategory,
-                                                      "connection to cached endpoints failed\n" +
-                                                      $"removing endpoints from cache and trying again\n{ex}");
+                            // Ignore the exception unless this is the last connector.
+                            // TODO retry with non cached endpoints
+                            if (ReferenceEquals(endpoint, last))
+                            {
+                                throw;
+                            }
                         }
-                        return await GetConnectionAsync(excludedConnectors, cancel).ConfigureAwait(false);
                     }
-                    throw;
-                }
-
-                if (CacheConnection)
-                {
-                    _connection = connection;
                 }
             }
+            finally
+            {
+                linkedCancellationSource.Dispose();
```

Comment:
Use `using var` instead of for the linkedCancellationSource.

Reply:
fixed